web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,439 @@
+        pushID,
+        state,
+        startTime,
+        endTime,
+        row: 0
+      } as Item);
+    });
+
+    // Assign a row value to each push representing their horizontal placement
+    // on the timeline. Each row index corresponds to one group.
+    this.divideIntoRows();
+  }
+
+  /**
+   * Divides the data into row by assigning a row index (corresponding
+   * to their row placement on the timeline) such that all intervals within a
+   * row do not overlap. The algorithm takes a greedy approach by sorting
+   * the row by increasing start time, sequentially picking the next interval,
+   * and removing all intervals it overlaps with. It thus fits as many intervals
+   * as possible in the first row, does so for each row until there are
+   * no more intervals left. The runtime is O(n * log(n)), which results in
+   * the case that all intervals are overlapping.
+   */
+  private divideIntoRows(): void {
+    let data: Item[] = [...this.data]; // Creates copy of this.data
+
+    data.sort((a, b) => {
+      return a.startTime - b.startTime;
+    });
+
+    let rowIndex = 0;
+    let overlappingIntervals = [this.data[0]]; // Initialized to arbitrary value to avoid premature return
+    while (overlappingIntervals.length !== 0) {
+      let lastEndTime = 0;
+      overlappingIntervals = [];
+      for (const interval of data) {
+        // To check for interval overlap with the already determined set of
+        // non-overlapping intervals, we simply need to check if the start time
+        // of the current interval is earlier than the last added end time.
+        // This is a consequence of all events being sorted by start time.
+        if (interval.startTime >= lastEndTime) {
+          const intervalInData = this.data.find(({ pushID }) => pushID === interval.pushID);
+          if (intervalInData) {
+            intervalInData.row = rowIndex;
+          }
+          lastEndTime = interval.endTime;
+        } else {
+          overlappingIntervals.push(interval);
+        }
+      }
+      rowIndex++;
+      lastEndTime = 0;
+      data = overlappingIntervals;
+    }
+    this.numRows = rowIndex;
+  }
+
+  /**
+   * Converts the date (in milliseconds) to a datetime string. The returned
+   * date will have a format of YYYY-MM-DD hh-mm, with single digits padded
+   * with a leading zero.
+   *
+   * @param d Holds one interval's data on the timeline.
+   */
+  private formatDate = (d: any) => {
+    const date = new Date(d);
+    const dateTimeFormat = new Intl.DateTimeFormat('en', {
+      year: 'numeric',
+      month: '2-digit',
+      day: '2-digit',
+      hour: '2-digit',
+      minute: '2-digit'
+    });
+
+    const formatToParts = 'formatToParts';
+    const [{ value: month }, ,
+           { value: day }, ,
+           { value: year }, ,
+           { value: hour }, ,
+           { value: minute }]
+      = dateTimeFormat[formatToParts](date);
+    
+    return `${year}-${month}-${day} ${hour}:${minute}`;
+  }
```

Comment:
Can we use `formatDate` provided by `@angular/common`?

Example: https://github.com/googleinterns/step189-2020/blob/main/web/src/app/pipes/date-nsec.pipe.ts